fmp_server.py

```diff
@@ -305,18 +305,51 @@ def recvReqLayerUp(self, packet_id, data, seq):
          - Relocate > Select Server
         """
         if self.session.layer == 2:
-            city = self.session.get_city()
-            if city.leader == self.session:
-                if len(city.players) > 1:
-                    # TODO: Transfer the leadership to another player
-                    pass
-                city.leader = None
+            new_leader = self.session.try_transfer_city_leadership()
+            if new_leader is not None:
+                self.sendNtcLayerHost(new_leader, seq)
 
             ntc_data = pati.lp2_string(self.session.capcom_id)
             self.server.layer_broadcast(self.session, PatID4.NtcLayerOut,
                                         ntc_data, seq)
         self.sendAnsLayerUp(data, seq)
 
+    def recvReqLayerEnd(self, packet_id, data, seq):
+        """ReqLayerEnd packet.
+
+        ID: 64020100
+        JP: レイヤ終了要求
+        TR: Layer end request
+        """
+        if self.session.layer > 1:
```

Comment:
I'd like to avoid exposing the `layer` implementation logic to the server network code. A proper method should be created in `session` like `session.leave_layer()`.

Ideally, that'd reduce this method code to:
```python
new_host = self.session.try_transfer_city_leadership()
if new_host:
    self.sendNtcLayerHost(new_host, seq)
self.session.leave_layer()  # You might need to pass the server or self as parameter for the broadcast, I'll have to figure out a better design for this.
self.sendAnsLayerEnd(seq)
```
I'm also open to suggestions regarding other design possibilities.

Reply:
You can also move the logic in the `session.layer_end` method if it's more fitting.

Reply:
When I coded this I was valuing keeping the `layer_broadcast` in the server network code over keeping the `layer` implementation logic out of it. It came down to doing it one way or the other, since the logic of `layer_broadcast` (for both `NtcLayerHost` and `NtcLayerOut`) has to be done *before*  the server removes the player from the city, but *after* the server checks to see if the player is in a city.

I did pack a chunk of the logic into `session` already in the form of `try_transfer_city_ownership`, which notifies the `server` that packets need to be sent only if the layer changed hands to a new host.

So, I could move the logic into `session`, but I think it would require that I also move a lot of packet-sending logic in there as well, which there currently isn't any.

Reply:
Feel free to disregard my point for now. I'll try to improve the current session abstraction. Basically, my initial thought was to separate the codebase like this:
 - session module: holds the current player network state.
 - database module: database abstraction interface.
 - server module: server module that will inherit protocol implementation or implement it.

I initially planned to implement the broadcast method into the session class but my implementation wasn't ready for that. So I used the server class instead.